src/libraries/System.Net.Sockets/tests/FunctionalTests/LocalEndPointTest.cs

```diff
@@ -0,0 +1,260 @@
+                Assert.Equal(clientPortAfterBind, ((IPEndPoint)receiveFromEP).Port);
+
+                IAsyncResult sendToResult = client.BeginSendTo(new byte[] { 4, 5, 6 }, 0, 3, SocketFlags.None, sendToEP, null, null);
+                sendToResult.AsyncWaitHandle.WaitOne();
+                client.EndSendTo(sendToResult);
+
+                Assert.Equal(Wildcard, GetLocalEPAddress(client)); // stays as wildcard after async WSASendTo
+                Assert.Equal(clientPortAfterBind, GetLocalEPPort(client));
+
+                buf = new byte[3];
+                receiveFromEP = new IPEndPoint(Wildcard, 0);
+                server.ReceiveFrom(buf, ref receiveFromEP);
+
+                Assert.Equal(new byte[] { 4, 5, 6 }, buf);
+                Assert.Equal(Loopback, ((IPEndPoint)receiveFromEP).Address); // received from specific address
+                Assert.Equal(clientPortAfterBind, ((IPEndPoint)receiveFromEP).Port);
+            }
+        }
+
+        [Fact]
+        public void UdpSocket_ClientNotBound_LocalEPBecomesWildcardOnSendTo()
+        {
+            using (Socket server = CreateUdpSocket())
+            using (Socket client = CreateUdpSocket())
+            {
+                int serverPort = server.BindToAnonymousPort(Wildcard);
+
+                Assert.Null(client.LocalEndPoint); // null before sendto
+
+                var sendToEP = new IPEndPoint(Loopback, serverPort);
+
+                client.SendTo(new byte[] { 1, 2, 3 }, sendToEP);
+
+                Assert.Equal(Wildcard, GetLocalEPAddress(client)); // wildcard after sendto
+
+                byte[] buf = new byte[3];
+                EndPoint receiveFromEP = new IPEndPoint(Wildcard, 0);
+                server.ReceiveFrom(buf, ref receiveFromEP);
+
+                Assert.Equal(new byte[] { 1, 2, 3 }, buf);
+            }
+        }
+
+        [Fact]
+        public void UdpSocket_ClientNotBound_LocalEPBecomesWildcardOnAsyncSendTo()
+        {
+            using (Socket server = CreateUdpSocket())
+            using (Socket client = CreateUdpSocket())
+            {
+                int serverPort = server.BindToAnonymousPort(Wildcard);
+
+                Assert.Null(client.LocalEndPoint); // null before async WSASendTo
+
+                var sendToEP = new IPEndPoint(Loopback, serverPort);
+
+                IAsyncResult sendToResult = client.BeginSendTo(new byte[] { 4, 5, 6 }, 0, 3, SocketFlags.None, sendToEP, null, null);
+                sendToResult.AsyncWaitHandle.WaitOne();
+                client.EndSendTo(sendToResult);
+
+                Assert.Equal(Wildcard, GetLocalEPAddress(client)); // wildcard after async WSASendTo
+
+                byte[] buf = new byte[3];
+                EndPoint receiveFromEP = new IPEndPoint(Wildcard, 0);
+                server.ReceiveFrom(buf, ref receiveFromEP);
+
+                Assert.Equal(new byte[] { 4, 5, 6 }, buf);
+            }
+        }
```

Comment:
These 2 tests seem to test the same thing for two different connect implementations. I suggest dedupe their code and also cover the remaining ones: `ConnectAsync(SocketAsyncEventArgs)`  and `SocketTaskExtensions.ConnectAsync(this Socket, ....)`. For this the best would be to utilize  `SocketTestHelperBase<T>`.

Reply:
Rewrote the tests, please check that I understood you correctly :)